programs/bpf_loader/src/lib.rs

```diff
@@ -440,20 +441,88 @@ fn process_instruction_common(
+            } else {
+                instruction_context.get_index_of_instruction_account_in_transaction(
+                    index_in_instruction
+                        .saturating_sub(instruction_context.get_number_of_program_accounts()),
+                )
+            }
+        }
+
+        fn try_borrow_account<'a>(
+            transaction_context: &'a TransactionContext,
+            instruction_context: &'a InstructionContext,
+            index_in_instruction: IndexOfAccount,
+        ) -> Result<BorrowedAccount<'a>, InstructionError> {
+            if index_in_instruction < instruction_context.get_number_of_program_accounts() {
+                instruction_context
+                    .try_borrow_program_account(transaction_context, index_in_instruction)
+            } else {
+                instruction_context.try_borrow_instruction_account(
+                    transaction_context,
+                    index_in_instruction
+                        .saturating_sub(instruction_context.get_number_of_program_accounts()),
+                )
+            }
+        }
+
+        let first_instruction_account = {
+            let borrowed_root_account =
+                instruction_context.try_borrow_program_account(transaction_context, 0)?;
+            let owner_id = borrowed_root_account.get_owner();
+            if native_loader::check_id(owner_id) {
+                1
+            } else {
+                0
+            }
+        };
+        let first_account_key = transaction_context.get_key_of_account_at_index(
+            get_index_in_transaction(instruction_context, first_instruction_account)?,
+        )?;
+        let second_account_key = get_index_in_transaction(
+            instruction_context,
+            first_instruction_account.saturating_add(1),
+        )
+        .and_then(|index_in_transaction| {
+            transaction_context.get_key_of_account_at_index(index_in_transaction)
+        });
+        let program_id = instruction_context.get_last_program_key(transaction_context)?;
+        if first_account_key == program_id
+            || second_account_key
+                .map(|key| key == program_id)
+                .unwrap_or(false)
+        {
+        } else {
+            let first_account = try_borrow_account(
+                transaction_context,
+                instruction_context,
+                first_instruction_account,
+            )?;
+            if first_account.is_executable() {
+                ic_logger_msg!(log_collector, "BPF loader is executable");
+                return Err(InstructionError::IncorrectProgramId);
+            }
+        }
+    }
+
     let program_account =
         instruction_context.try_borrow_last_program_account(transaction_context)?;
 
     // Program Management Instruction
     if native_loader::check_id(program_account.get_owner()) {
         drop(program_account);
-        if instruction_context
```

Comment:
Do we want to add this code to the else clause of the feature gate check? Or is this check no longer useful after feature activation?

Reply:
This is no longer needed, it was originally intended to replace the old code for throwing the error.
But now that we are restoring the old code, we don't need this replacement anymore.